### ISSUE TITLE
Add and use hash-canonicalizing field comparator [phab:D1212]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1586]
_Created by **shahms** at 2017-01-03 21:01:56_: 

Next step in cleaning up shell-based extractor tests

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-03 22:46:49_:

A brief comment would be helpful (or a rename to `FindNestedFieldByLowercasePath` or similar).

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2017-01-03 22:46:49_:

It's not clear to me why you're assigning new `hash###` strings. IIRC in the original tests the `hash###` tokens were used in a pseudo-unification algorithm with no backtracking. Here, you assign, e.g.:

```
CHash(hash0) => hash1
CHash(23d6e5ea4f1f8cf7f2b1abad9a48bc92a2d97411d45182b8a50baf36333de713) => hash1
CHash(hash1) => hash2
CHash(7222c88b57d041af3b9e998a483dccc44d15950bde1fb8df4ccb4b7dd7250962) => hash2
CHash(hash1) => hash2
CHash(7222c88b57d041af3b9e998a483dccc44d15950bde1fb8df4ccb4b7dd7250962) => hash2
```

I believe this will work, but it's surprising enough to comment.

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2017-01-03 22:46:49_

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2017-01-03 23:18:54_:

This is semantically equivalent to the original (from `kindex_tool_main.cc`), except the ordinal is increased by one.  However, as I don't need to use strings at all, I've switched back to using `size_t` and added a comment.

**Comment by kythe** [comment]
_**shahms** wrote at 2017-01-03 23:19:01_:

  - Use insertion-index rather than generated hash string for comparison
  - Rename and add a comment to FindNestedFieldByLowercasePath

**Comment by kythe** [comment]
_**schroederc** wrote at 2017-08-17 17:14:30_:

What's the status of this?

**Comment by kythe** [comment]
_**shahms** wrote at 2017-08-17 17:18:49_:

>>! In (earlier comment), schroederc wrote:
> What's the status of this?

On the back burner, potentially abandoned.  Certainly low priority.,

**Comment by kythe** [comment]
_**jrtom** wrote at 2017-10-18 16:48:57_:

Echoing schroederc's question from a couple of months ago: is this worth pursuing further, or should we just abandon it?  It looks like it was already approved a while back; what further action is needed?

**Comment by kythe** [comment]
_**shahms** wrote at 2017-10-18 16:58:35_:

  - Update newest allowed Bazel version

**Comment by kythe** [comment]
_**shahms** closed via commit 811b9e780018a94658e97aa373df78db1fbfd82e at 2017-10-18 17:00:32_
